File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    static vector<int> inorderTraversal(const TreeNode* root);
};

vector<int> Solution::inorderTraversal(const TreeNode* root)
{
    vector<int> ret;

    for (stack<const TreeNode*> nodes; root || !nodes.empty();
                                       root = root->right) {
        for (; root; root = root->left) nodes.push(root); // traverse left

        root = nodes.top();
        nodes.pop();
        ret.push_back(root->val);
    }

    return ret;
}

File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> InorderTraversal(TreeNode root) {
        var ret = new List<int>();
        var stack = new Stack<TreeNode>();
        
        while (root != null || stack.Count != 0) {
            // Go left as far as possible.
            for (; root != null; root = root.left) stack.Push(root);
            
            // Process the current node.
            var cur = stack.Pop();
            ret.Add(cur.val);
            
            // Go right by one (if we can).
            root = cur.right;
        }
        
        
        return ret;
    }
}
